### ISSUE TITLE
[Feature] Mouse helper (tracking mouse position, clicks, drag & drops, selections)

### ISSUE THREAD

**Issue by thernstig** [microsoft/playwright#1374]
It would be nice to natively have support for a mouse helper. So if running in headful mode, or taking a screenshot, or video recording, one would see where the mouse is, text selections, drag & drops, and where it clicks during a failure.

@aslushnikov already had one in Puppeteer but it required installing it, but maybe it could be default-installed on the browserContext or page instead if wanted, or even for the browser. Having it out of the box would improve usability for all users. Especially during debugging.

https://github.com/puppeteer/puppeteer/issues/4378#issuecomment-499726973

**Comment by alexsorinsky** [comment]
very important feature +1

**Comment by aslushnikov** [comment]
This will be addressed by https://github.com/microsoft/playwright/pull/3526#pullrequestreview-472161209

**Comment by LanderBeeuwsaert** [comment]
Yes, in our test cases a lot of the time it's difficult to know at exactly what line playwright failed. Having visual confirmation; most importantly in the video if & where mouse clicks were performed will for some issues certainly help us to identify them. In cypress when you hover over an action in the action bar & the action is a click, you see the click on the screenshot for that moment, which is quite helpful.

**Comment by aslushnikov** [comment]
This should be now addressed with a much more advanced ui - please check out the new playwright inspector: https://playwright.dev/docs/inspector/

I'll close this bug now. We continue working on making inspector UI nicer and more helpful, so if you have any issues or suggestions for it, please file separate issues!

**Comment by thernstig** [comment]
@aslushnikov Maybe I am missing something, but how does this work if I execute a `page.screenshot()` as a custom handling when failures happen? Or if I use the record functionality, to see the mouse move?

I.e. I want this to use this in CI when a test case fails, I take a screenshot, and see where it failed. Without having to re-run with the inspector.

**Comment by LanderBeeuwsaert** [comment]
yeah, that would also be our preferred usage. We want to run it during our CI runs. We found out that +- 40% of the work is building the tests; but +-60% of the time we spend on hardening them when 1 in 20 runs a command fails because for that run it behaves a bit different because the CPU was slower or whatever. It's there that we would like to gain time.

**Comment by aslushnikov** [comment]
@thernstig @LanderBeeuwsaert good points! Re-opening then

**Comment by thernstig** [comment]
@aslushnikov great :) No idea of how you achieved this with the inspector, but hopefully you can re-use some of that code for also this scenario we described i.e. in CI.

**Comment by klhex** [comment]
Interestingly, there is already one available separately and can be imported into Playwright: https://gist.github.com/aslushnikov/94108a4094532c7752135c42e12a00eb

However, of course it would be really nice and helpful, to have such a feature integrated in Playwright, including the possibility to activate it as needed via API (e.g. ```page.debug( { mousehelper: true } )```).

**Comment by klhex** [comment]
@thernstig Right, I didn't notice that it was already linked in the first comment. Guess it's time to call it a day. ;) (However, I did notice that it was written by aslushnikov. :)

**Comment by ChrisLasar** [comment]
Any news on this? Would love to see this feature. The maybe related pull request (https://github.com/microsoft/playwright/pull/3526#pullrequestreview-472161209) is still closed. Would like to see this feature to automatize generation of small screen videos for tutorials. Any other ideas and hints on how to do this?

**Comment by lourd** [comment]
Linking to related issues #6737 and #5816. These are critical for my team for using Playwright to test our WebGL app, https://spatial.io

**Comment by klhex** [comment]
This would be especially useful for CI/CD environments! In my case, I am running all my tests in the background (CI/CD) in headed mode with screen recording enabled by default, incl. automatic upload of the recordings to S3. With a mouse helper that I would also activate by default for all my tests, I would be able to detect issues related to flakiness, most prominent example being that PW confirms it has successfully clicked on "Next" button but nevertheless the application is not navigating to the next page, as it normally should. If I run the test again, usually the issue does not reappear (flake) and therefore I cannot use the PW trace viewer to analyze the issue further. Currently, the only alternative would be to enable the trace viewer by default for all test executions but this would be overkill since such flake issues only occur rarely and having it always activated would slow down the tests quite a bit...

**Comment by baptisteArno** [comment]
+1, that would be so helpful!

**Comment by DmitryLobachov** [comment]
+1

**Comment by ivelin** [comment]
Seems like using Tracing with Action Snapshot turned on does the job:
https://playwright.dev/docs/trace-viewer#snapshots

**Comment by cenfun** [comment]
You can simulate a mouse with [mouse-helper](https://github.com/cenfun/mouse-helper)

```sh
npm i mouse-helper
```

Usage 
```js
const { chromium } = require('playwright');

const browser = await chromium.launch();

const context = await browser.newContext();
await context.addInitScript({
    path: './node_modules/mouse-helper/dist/mouse-helper.js'
});

const page = await context.newPage();
await page.goto("your page url");

await page.evaluate(() => {
     window['mouse-helper']();
});

await page.mouse.move(20, 50);
await page.screenshot({
    path: "your screenshot path"
});

await page.mouse.down();
await page.screenshot({
    path: "your screenshot path"
});

await page.mouse.up();

```
![image](https://github.com/microsoft/playwright/assets/334577/3db876c4-212d-47da-8f5d-ea860199bb4b)

**Comment by storenth** [comment]
I think trace (and its viewer) doesn't cover all dev needs. Looks like some helper need to draw what is mouse/keyboard doing as part of the video object. Second vector is a Chrome extension to highlight mouse events.

**Comment by fjcomas** [comment]
Is there a way to add the red circle (tracer) to the videos? it does it for the Actions but when a video is created it loses it and it is hard to follow.

**Comment by ajscott-hcs** [comment]
I'd like to add to this conversation where it would be nice to have the ability to have the cursor within the test, for screenshot or video purposes - that's not in a debug or trace viewer context.

This [review here](https://github.com/microsoft/playwright/issues/5816) was exactly what I personally needed for my project but it was closed

**Comment by hriverahdez** [comment]
I tried the solution posted above by @cenfun using the mouse-helper and that indeed helps. However, a solution like this also adds the mouse cursor itself to the snapshot, so a slightly different cursor position leads to a failed test (ofc we could also add a pixel diff threshold). 

It would be nice however if Playwright itself provided such option which would get excluded from screenshot matching. Our use cases is that we have a webgl app and we want to fire off actions based on canvas positions and run snapshot assertions and in many cases having a  visible mouse cursor helps to sanity-check the snapshot and for easier debugging

**Comment by scholtz** [comment]
strange that this is not implemented in core library after 6 years this ticket is open. the use case is quite obvious - programmatically create videos for educational purposes using playwright.. mouse MUST be visible in such videos. cypress does this better now

**Comment by prakersh** [comment]
For a different take on visually tracking interactions, [onUI](https://github.com/onllm-dev/onUI) is a free, open-source browser extension with a freehand draw mode that overlays on any webpage. It lets you mark up interactions directly in the browser, and the annotations are accessible programmatically via its local MCP server.

**Comment by yury-s** [comment]
Playwright now renders action annotations natively:

```js
await page.screencast.showActions({ position: 'top-right' });
```

This highlights the interacted element, shows the action title and draws a cursor that animates between action points — in headed mode and in recorded videos. In the test runner:

```js
export default defineConfig({
  use: {
    video: {
      mode: 'on',
      show: { actions: { position: 'top-left' } },
    },
  },
});
```

The same is available on the `recordVideo` context option as `showActions`. For tutorial-style recordings there are also chapter titles and custom HTML overlays, see https://playwright.dev/docs/videos and https://playwright.dev/docs/api/class-screencast.
